src/modes/edit_message.cc

```diff
@@ -719,15 +738,50 @@ namespace Astroid {
       account_no ++;
     }
 
+    update_message_id_rhs_from_sender(a->email);
```

Comment:
This might cause problems if this is work on an existing draft which exists in notmuch since the message is identified in notmuch by its ID. In that case the message would have to be move/copied in notmuch.

Reply:
This only happens for drafts which have not been sent. After a message has been sent, the message id is never touched again, of course.

For as long as I keep editing it within astroid (and maybe an external editor), the message will only exist as a tempfile in `~/.cache/astroid/`, and everything works fine. The sent message is filed in the sent maildir, but gets the message id as it's filename.

If I save the draft before sending, and change the message id, astroid crashes with a message of "[error] db: could not find message". So it seems notmuch has a built-in assumption that the filename is identical to the message id (and never checks the message id inside the file again).

So I **think** changing the message id **could** work if I manage to do three things:
1. use the real message id as the filename when drafting a message
2. update the file name when the message id is updated
3. have notmuch rescan the the files if and when the filename of the draft has been updated

I'll make a few experiments and keep you posted.

Reply:
Ok, I got as far as decoupling all file name generation (including for storing in `MAILDIR`) from the message id. It seems to not cause any major mayhem, except one small quirk (but which shows up in the debug log only). See last commit 7b76ff66f5bdd076483eb1590c511ca9527f05a5.

The test run is as follows:
1. configure astroid to use an external editor, and start composing a new message
2. when back in astroid, save the new message as a draft
3. revisit the draft, and edit it once again in the external editor
4. in the external editor, change the `From:` address to another account of yours
5. when back in astroid, observe how the from address drop-down button has changed and shows the new account
6. send the message without further ado
7. dig into `MAILDIR/sent/cur` (or whtever you've configured it to) to discover the message id of the message you just sent
8. return to astroid and search for the message of the message just sent (`o`, `id:...`); there will be a search result

So notmuch seems to pick up the new message fine.

Where the cogwheels start making grinding noises in the debug log is in/around `on_message.cc`. Right after successful sending, `AddSentMessage::emit` is called with the **new** message id. This is followed by three db search operations. Then, `OnMessageAction::emit` is called, but with the **old** message id. Most of the times, nothing else bad happens, but occasionally it causes a crash somewhere in glibmm when searching for the thread relating to the old message id.

It would thus seem that somewhere in/around `on_message.cc` I would need to check whether the message id has been updated, and take note of it in the appropriate member variable. As mentioned above, notmuch already seems to grasp the new message id.

Reply:
Notmuch doesn't pick up anything automatically. It either has to be be set in a consistent state by _database_remove/_database_index calls or `notmuch new` must be run. We should not rely on the latter, and the database should always be in a consistent state after astroid has done any modifications. The `OnMessage` error should probably not be fixed in `OnMessage`, it likely stems from some out-of-date message-id somewhere else and should be fixed there (since previously it was safe to assume that the message-id was immutable).

Reply:
> Notmuch doesn't pick up anything automatically.

I know. So it seems that the existing code updates the db often enough already?

Reply:
> The OnMessage error should probably not be fixed in OnMessage, it likely stems from some out-of-date message-id somewhere else and should be fixed there (since previously it was safe to assume that the message-id was immutable).

I am currently trying to figure out where this somewhere else could be.

I have added a `LOG` statement in `void Message::on_message_updated (Db * db, ustring _mid)` which maybe can shed a little more light. The log excerpt is below. The first "emitted updated signal for message:" is followed by three invocations of `Message::on_message_updated`, where the last one features the new message id. The second "emitted updated signal for message:" is followed by two invocations of `Message::on_message_updated`, but before the third invocation occurs, there is a "db: could not find message:", followed by a single Message::on_message_updated` with the old message id. Then it goes on to crash.

```
[15:27:30.060336] [0x00007000093a8000] [debug] db: open time: 94.149 ms.
[15:27:30.060437] [0x00007000093a8000] [info] db: adding sent message: /Users/alexanderadolf/Library/Mail/Maildir/sent/cur/5si6iga31o0m6fdo9ablfyd92akuhus8yzilnk5fj7ogu6pjee:2,
[15:27:30.071548] [0x00007000093a8000] [info] db: closing db.
[15:27:30.091940] [0x00007000093a8000] [debug] db: rw: releasing lock.
[15:27:30.091991] [0x00007000093a8000] [debug] db: rw-s: releasing lock.
[15:27:30.092027] [0x00007000093a8000] [info] db: open db read-write.
[15:27:30.092047] [0x00007000093a8000] [debug] db: rw-s: waiting for rw lock.. (r-o open: 0)
[15:27:30.092100] [0x00007000093a8000] [debug] db: rw-s lock acquired.
[15:27:30.191504] [0x00007000093a8000] [debug] db: open time: 110.952 ms.
[15:27:30.195270] [0x00007000093a8000] [info] db: closing db.
[15:27:30.215011] [0x00007000093a8000] [debug] db: rw: releasing lock.
[15:27:30.215063] [0x00007000093a8000] [debug] db: rw-s: releasing lock.
[15:27:30.262033] [0x00007fffad7a0340] [info] db: open db read-only, waiting for lock..
[15:27:30.262096] [0x00007fffad7a0340] [debug] db: read-only got lock.
[15:27:30.262355] [0x00007fffad7a0340] [debug] db: open time: 0.331 ms.
[15:27:30.262401] [0x00007fffad7a0340] [info] actions: emitted updated signal for message: I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de
[15:27:30.262441] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de
[15:27:30.262474] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de
[15:27:30.262507] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de
[15:27:30.262592] [0x00007fffad7a0340] [info] actions: emitted changed signal for thread: 0000000000000296
[15:27:30.262621] [0x00007fffad7a0340] [debug] searches: refreshing..
[15:27:30.262643] [0x00007fffad7a0340] [debug] searches: skipping, not visible.
[15:27:30.262659] [0x00007fffad7a0340] [info] ql (0): tag:inbox, got changed thread signal: 0000000000000296
[15:27:30.262712] [0x00007fffad7a0340] [debug] db: checking if thread: 0000000000000296 matches query: tag:inbox
[15:27:30.262825] [0x00007fffad7a0340] [debug] db: thread in query check: 0.12 ms.
[15:27:30.262850] [0x00007fffad7a0340] [debug] ql: updated: did not find thread, time used: 0.177 ms.
[15:27:30.262867] [0x00007fffad7a0340] [info] ql (1): tag:draft, got changed thread signal: 0000000000000296
[15:27:30.262924] [0x00007fffad7a0340] [debug] db: checking if thread: 0000000000000296 matches query: tag:draft
[15:27:30.268751] [0x00007fffad7a0340] [debug] db: thread in query check: 0.689 ms.
[15:27:30.268903] [0x00007fffad7a0340] [debug] ql: updated: did not find thread, time used: 0.864 ms.
[15:27:30.268971] [0x00007fffad7a0340] [info] ql (2): tag:draft, got changed thread signal: 0000000000000296
[15:27:30.269106] [0x00007fffad7a0340] [debug] db: checking if thread: 0000000000000296 matches query: tag:draft
[15:27:30.269405] [0x00007fffad7a0340] [debug] db: thread in query check: 0.316 ms.
[15:27:30.269468] [0x00007fffad7a0340] [debug] ql: updated: did not find thread, time used: 0.454 ms.
[15:27:30.269539] [0x00007fffad7a0340] [info] db: closing db.
[15:27:30.269733] [0x00007fffad7a0340] [debug] db: ro: waiting for lock to close..
[15:27:30.269789] [0x00007fffad7a0340] [debug] db: ro: closing..
[15:27:30.269869] [0x00007fffad7a0340] [info] db: open db read-only, waiting for lock..
[15:27:30.269912] [0x00007fffad7a0340] [debug] db: read-only got lock.
[15:27:30.270327] [0x00007fffad7a0340] [debug] db: open time: 0.441 ms.
[15:27:30.270410] [0x00007fffad7a0340] [info] actions: emitted updated signal for message: I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com
[15:27:30.270481] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com
[15:27:30.270543] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com
[15:27:30.270751] [0x00007fffad7a0340] [error] db: could not find message: I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com, status: 0
[15:27:30.270827] [0x00007fffad7a0340] [info] message: emitted changed signal for message: I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com: 0
[15:27:30.270884] [0x00007fffad7a0340] [debug] msg: on_message_updated: mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@mnet-online.de / _mid = I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com
[15:27:30.270961] [0x00007fffad7a0340] [error] db: could not find message: I%9Xu/8c8wy2'Xz%|~MKA8urvpiK}187kV/+fHW^@condition-alpha.com, status: 0
[15:27:30.271047] [0x00007fffad7a0340] [info] actions: emitted updated and changed signal for thread: 0000000000000295
[15:27:30.271341] [0x00007fffad7a0340] [error] db: could not find thread: 0000000000000295
[15:27:30.271480] [0x00007fffad7a0340] [info] db: closing db.
[15:27:30.271575] [0x00007fffad7a0340] [debug] db: ro: waiting for lock to close..
[15:27:30.271612] [0x00007fffad7a0340] [debug] db: ro: closing..

(<unknown>:37441): glibmm-ERROR **:
unhandled exception (type std::exception) in signal handler:
what: db: could not find thread!

Trace/BPT trap: 5
```

When running it in the debugger, the stack backtrace is this:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=EXC_I386_BPT, subcode=0x0)
  * frame #0: 0x00000001031e2f17 libglib-2.0.0.dylib`_g_log_abort + 22
    frame #1: 0x00000001031e2da2 libglib-2.0.0.dylib`g_log_default_handler + 206
    frame #2: 0x00000001031e2a5e libglib-2.0.0.dylib`g_logv + 878
    frame #3: 0x00000001031e26d2 libglib-2.0.0.dylib`g_log + 120
    frame #4: 0x000000010393c6cd libglibmm-2.4.1.dylib`Glib::exception_handlers_invoke() + 303
    frame #5: 0x000000010393b064 libglibmm-2.4.1.dylib`Glib::DispatchNotifier::pipe_io_handler(Glib::IOCondition) + 326
    frame #6: 0x000000010393f13c libglibmm-2.4.1.dylib`Glib::IOSource::dispatch(sigc::slot_base*) + 52
    frame #7: 0x000000010393e253 libglibmm-2.4.1.dylib`Glib::Source::dispatch_vfunc(_GSource*, int (*)(void*), void*) + 39
    frame #8: 0x00000001031dce64 libglib-2.0.0.dylib`g_main_context_dispatch + 258
    frame #9: 0x00000001031dd162 libglib-2.0.0.dylib`g_main_context_iterate + 426
    frame #10: 0x00000001031dd1ba libglib-2.0.0.dylib`g_main_context_iteration + 55
    frame #11: 0x0000000102bafc56 libgio-2.0.0.dylib`g_application_run + 468
    frame #12: 0x000000010001e839 astroid`Astroid::Astroid::run(this=0x0000000106d00480, argc=1, argv=0x00007ffeefbff530) at astroid.cc:320
    frame #13: 0x000000010000d5aa astroid`main(argc=1, argv=0x00007ffeefbff530) at main.cc:5
    frame #14: 0x00007fff74150145 libdyld.dylib`start + 1
```

Does any of the above ring any bells with you where I should be taking a closer look?